src/pymor/reductors/basic.py

```diff
@@ -8,197 +8,249 @@
             return self._last_rom
         else:
-            return self._reduce_to_subbasis(dim)
+            return self._reduce_to_subbasis(dims)
 
     def _reduce(self):
+        projected_operators = self.project_operators()
+        rom = self.build_rom(projected_operators)
+        rom = rom.with_(name=f'{self.fom.name}_reduced')
+        rom.disable_logging()
+        return rom
 
-        fom = self.fom
-        RB = self.RB
+    def _reduce_to_subbasis(self, dims):
+        projected_operators = self.project_operators_to_subbasis(dims)
+        rom = self.build_rom(projected_operators)
+        rom = rom.with_(name=f'{self.fom.name}_reduced')
+        rom.disable_logging()
+        return rom
 
-        if any(k in self.vector_ranged_operators for k in fom.operators) and not self.basis_is_orthonormal:
-            projection_matrix = RB.inner(RB, self.product)
-            projection_op = NumpyMatrixOperator(projection_matrix, source_id=RB.space.id, range_id=RB.space.id)
-            inverse_projection_op = InverseOperator(projection_op, 'inverse_projection_op')
+    @abstractmethod
+    def project_operators(self):
+        pass
 
-        def project_operator(k, op):
-            if op is self.product and self.basis_is_orthonormal:
-                return IdentityOperator(NumpyVectorSpace(len(RB), RB.space.id), name=op.name)
-            elif k in self.vector_ranged_operators:
-                assert RB in op.range
-                pop = project(op, range_basis=RB, source_basis=RB if RB in op.source else None, product=self.product)
-                if not self.basis_is_orthonormal:
-                    return Concatenation([inverse_projection_op, pop])
-                else:
-                    return pop
-            else:
-                return project(op,
-                               range_basis=RB if RB in op.range else None,
-                               source_basis=RB if RB in op.source else None)
+    @abstractmethod
+    def build_rom(self, projected_operators):
+        pass
 
-        projected_operators = {k: project_operator(k, op) if op else None for k, op in fom.operators.items()}
+    def project_operators_to_subbasis(self, dims):
+        raise NotImplementedError
 
-        projected_products = {k: project_operator(k, p) for k, p in fom.products.items()}
+    def reconstruct(self, u, basis):
+        """Reconstruct high-dimensional vector from reduced vector `u`."""
+        return self.bases[basis][:u.dim].lincomb(u.to_numpy())
 
-        rom = fom.with_(operators=projected_operators, products=projected_products,
-                        visualizer=None, estimator=None,
-                        cache_region=None, name=fom.name + '_reduced')
-        rom.disable_logging()
+    def extend_basis(self, U, basis, method='gram_schmidt', pod_modes=1, pod_orthonormalize=True,
+                     check_orthonormality=None, check_tol=None, copy_U=True):
+        if check_orthonormality is None:
```

Comment:
Is it intentional to leave ```check_orthonormality``` potentially unset below?
Should this be ```check_orthonormality = check_orthonormality or (basis in self.products)```?

Reply:
There is a default for `check_orthonormality` in the free `extend_basis` function.